### PR TITLE
Allow library features to be marked internal as well

### PR DESCRIPTION
**Pull request by compiler-errors**
allows `unstable(..., is_internal, ...)` to mark a library feature as internal. Not sure if we should just have a different stability attribute, like `#[internal(...)]` with the same fields as `unstable`.

Part of #115597.

This does no modification to the standard library yet.

### CODE REVIEW

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -1744,13 +1744,10 @@ rustc_queries! {
...
-    query lib_features(_: ()) -> &'tcx LibFeatures {
-        arena_cache
-        desc { "calculating the lib features map" }
-    }
-    query defined_lib_features(_: CrateNum) -> &'tcx [(Symbol, Option<Symbol>)] {
+    query lib_features(_: CrateNum) -> &'tcx LibFeatures {
         desc { "calculating the lib features defined in a crate" }
         separate_provide_extern
+        arena_cache
     }
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/query/mod.rs:1751]: I consolidated this into one query -- it's not necessarily a requirement of this PR to do so, so if that's not desirable, I could revert it.

**Comment by petrochenkov** [comment]
Double-gating internal features was a mistake by itself, and now other unnecessary complexity is accumulating around that mistake.

File: compiler/rustc_passes/src/stability.rs [hunk 2]
```diff
@@ -1027,7 +1030,7 @@ pub fn check_unused_or_stable_features(tcx: TyCtxt<'_>) {
 
     // We always collect the lib features declared in the current crate, even if there are
     // no unknown features, because the collection also does feature attribute validation.
-    let local_defined_features = tcx.lib_features(()).to_vec();
+    let local_defined_features = tcx.lib_features(rustc_hir::def_id::LOCAL_CRATE);
```
Comment (cjgillot) [inline compiler/rustc_passes/src/stability.rs:1033]: Please import this.

File: compiler/rustc_middle/src/query/mod.rs (hunk 1 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/query/mod.rs:1751]: This is useful. Could you extract it in a standalone query? Or even a standalone PR, as there may be discussion on this PR.

File: compiler/rustc_lint/src/builtin.rs [hunk 3]
```diff
@@ -2278,6 +2278,21 @@ impl EarlyLintPass for IncompleteInternalFeatures {
...
+                    if *is_unstable {
+                        cx.emit_spanned_lint(
+                            INTERNAL_FEATURES,
+                            span,
+                            BuiltinInternalFeatures { name, note: None },
+                        );
+                    }
+                }
+            }
+        }
```
Comment (cjgillot) [inline compiler/rustc_lint/src/builtin.rs:2306]: Should all this method go in `rustc_passes::stability::check_unused_or_stable_features`?

**Comment by RalfJung** [comment]
I was thinking we could do this a lot simpler, and just trigger the lint for all features whose name ends in `_internal` or `_internals`. That should also work pretty well, shouldn't it? We'll have to rename `core_intrinsics` to `core_intrinsics_internal`, but other than that I think the internal library features already follow this convention.

**Comment by compiler-errors** [comment]
I kinda prefer the more general solution, but yeah, I'm happy doing it that way too.

**Pull request closed without merging**